### ISSUE TITLE
Make it easier to code gen the flutter <--> embedded golang bridge code

### ISSUE THREAD

**Issue by joe-getcouragenow** [jerson/flutter-openpgp#7]
Looks like you and others have the same intent as me :)

What do you think about this idea ??

If the API between the flutter layer and the embedded golang layer was a Protobuf, then we could code gen a ton of the code 
For example you need to hand code the dart <--> golang method channels, but this can easily be code generated by sniffing the protobufs later.

The relevant code touch points are here:
Flutter Native bridge:
https://github.com/jerson/flutter-openpgp/blob/master/lib/openpgp.dart#L14
Flutter Web bridge:
https://github.com/jerson/flutter-openpgp/blob/master/lib/web/openpgp_web.dart

**Comment by jerson** [comment]
I think it's an excellent idea, if you have some ideas on how to implement it it would help me a lot because it would allow me to expose more methods and better performance

**Comment by joe-getcouragenow** [comment]
Thanks for being receptive to the idea.  

SO assuming we make the protobuf the contract, then something like these:

https://github.com/emicklei/proto-contrib

https://github.com/moul/protoc-gen-gotemplate

From there we could gen the dart and goldn code using go templates

---

There is also FFI instead of protobufs, but i am a little wary of it being still a bit too new myself.

I saw that go-flutter seems to work with FFI: https://github.com/go-flutter-desktop/go-flutter/issues/58#issuecomment-530375169. 

---

SO Let me know what you think of this general approach..

**Comment by joe-getcouragenow** [comment]
Made some progress...
This helps to give the flutter web and flutter naive the same Isolate abstraction that you have now but in a cleaner way

https://github.com/deakjahn/flutter_isolate_web
It uses this https://pub.dev/packages/isolate_handler, but make its work for web.

It removes alot of the code here:
https://github.com/jerson/flutter-openpgp/blob/master/lib/web/openpgp_web.dart

**Comment by jerson** [comment]
@joe-getcouragenow im working now implementing with ffi i think is more appropiate but only will work for windows, linux , macos and hover, i really cant undersant how we could use with protobufs, maybe if you have an example flutter <- go will be great

**Comment by joe-getcouragenow** [comment]
@jerson Thanks for asking... This is complicated stuff for sure..

This is VERY interesting: https://github.com/dart-lang/sdk/issues/42982
It got merged into the Google VM FFI code !!

FFI will also work with android and ios ?
FFI is better and more best practice with Flutter, that method channel.
But it is perfectly possible to use FFI and Protobufs together and this is my main motivation in all this.

Here at the end you can see lots of devs decided to use Protobufs instead of FFI for integrating Dart and Golang together.
https://github.com/go-flutter-desktop/go-flutter/issues/334

Why use Protobuf:
Protobuf makes it easier to manage the Data Type aspects.
- So from architecture point of view Protobuf is the Model, and FFI is the Transport. The reason i give the metaphor of FFI being the transport is because they is not Service interface. Its like GRPC without the Service interface and only the Types.
I could not find a perfect example of this but this shows the idea:
https://github.com/mailgun/pebblezgo
- In their case zeromq is the Transport.
https://github.com/go-kit/kit
- they adhere to strict separation of the Types and the Transport also. as they say in the README "Pluggable serialisation and transport — not just JSON over HTTP"
- Because of this sep there are generators like this: https://github.com/sagikazarmark/protoc-gen-kit. And thats why i suggested that with Protobufs you can generate the Dart and Golang code for the stubs that use the Protobufs.
 

Hope the above makes sense.. Just ask if i did not explain it well enough.


Here is one example of what i mean. But they use Protobufs with MethodChannels, not FFI. Mostly because the code is old and FFI never worked back then. 
Its a bit hard to follow because they used a different embed approach on desktop versus mobile.


https://github.com/deromask

Their Proto contract
- https://github.com/deromask/dero-gui/tree/master/protos

Flutter calling into GO using Protos
- https://github.com/deromask/dero-gui/blob/master/lib/core/services/derocore.dart

Golang mobile core exposed to Flutter
https://github.com/deromask/derosuite/blob/master/cmd/mobile/core.go

Golang Desktop exposed to Flutter
https://github.com/deromask/dero-gui/blob/master/go/plugins/derocore/core.go#L56

---

**Comment by joe-getcouragenow** [comment]
@jerson 
no comments ? Curious what your thoughts are.

**Comment by jerson** [comment]
@hi @joe-getcouragenow actually sound like a great solution, i was looking all implementation and now i understand how to use protobuf, im gonna use in that way:

windows, linux,macos, hover: FFI + PROTOBUF
android, ios: gomobile + PROTOBUF

by now looks like ffi dont work well for mobile because i need to get the architecture in dart and i think gomobile is enough for that but in the future when ffi had a better implementation easly could be used to replace gomobile

im gonna base the implementation in derosuite repo he is using : 
android, ios: gomobile + PROTOBUF 
hover: go + PROTOBUF

**Comment by joe-getcouragenow** [comment]
Hey @jerson Thanks for the feedback. 

I have seen other people not using gomobile to compile for mobiles also. I suspect the golang team is hoping tinygo will take over the role of compiling f

Below is one example. I have seen others out there also, but will take a bit of googling.  
Also gio is being maintaining by a former maintainer of the gomobile code at google, and even he does not use it. 

I am not suggesting BTW that Flutter is crazy and gophers should drop it and use GIO. I am only highlighting that there are others ways of doing the compilation of golang to mobiles..
I have used GIO on mobiles and it works flawlessly in terms of compilation. 

https://gioui.org/
- supports Web, desktop and mobile.

Entry Point:
https://git.sr.ht/~eliasnaur/gio/tree/master/cmd/gogio/main.go#L172

Android
https://git.sr.ht/~eliasnaur/gio/tree/master/cmd/gogio/androidbuild.go#L201

IOS
https://git.sr.ht/~eliasnaur/gio/tree/master/cmd/gogio/iosbuild.go#L425

**Comment by joe-getcouragenow** [comment]
https://rogchap.com/2020/09/14/running-go-code-on-ios-and-android/

on mobile no need for gomobile

**Comment by jerson** [comment]
Wow nice, probable this is the missing part i was looking for
https://rogchap.com/2020/09/14/running-go-code-on-ios-and-android/

thanks @joe-getcouragenow , these days I have been a bit busy with work but I am taking this up next days

**Comment by jerson** [comment]
now I'm doing the go part before compilation and it looks like this, I think when it's finished it can be easily integrated with flutter

```go

// proxy
// Call ...
func Call(name string, payload []byte) ([]byte, error) {

	instance := NewInstance()
	var output proto.Message
	switch name {
	case "decrypt":
		output = instance.decrypt(payload)
	case "decrypt_bytes":
		output = instance.decryptBytes(payload)

	}

	return proto.Marshal(output)
}


////////////////////////////////////////


//typedef struct  { void* message; int size; char* error; } BytesReturn;
import "C"

// .... bindings for ffi
func Call(name *C.char,payload unsafe.Pointer, payloadSize C.int) *C.BytesReturn {
	output := (*C.BytesReturn)(C.malloc(C.size_t(C.sizeof_BytesReturn)))
	defer C.free(unsafe.Pointer(name))
	defer C.free(payload)

	result, err := bridge.Call(C.GoString(name),C.GoBytes(payload, payloadSize))
	if err != nil {
		output.error = C.CString(err.Error())
		return output
	}
	output.error = nil
	output.message = C.CBytes(result)
	output.size = C.int(len(result))
	return output
}

```

**Comment by jerson** [comment]
Its funny but now looks like libopenpgp_bridge(binding bridge)  is bigger than libgojni(gomobile) but i think probably is something that im missing :D , im gonna continue with this
![image](https://user-images.githubusercontent.com/325780/94092617-d702ad80-fde0-11ea-8746-02bc2a5a34b4.png)

**Comment by jerson** [comment]
![image](https://user-images.githubusercontent.com/325780/94096639-331dff80-fdea-11ea-8325-c8e4c0e5ddf3.png)
nvm i was building wrong, now has the same size or a bit less :D

**Comment by jerson** [comment]
hi @joe-getcouragenow  a couple of minutes ago i released v1.0.0-rc1, this is the first version that uses ffi for most platforms, i think it's a good start for the future to add method in an easy way, thanks for your help, i think i couldn't do this without your help, the next step is to make web async


windows: ffi + protobuf
linux: ffi + protobuf
android: ffi + protobuf
macos: ffi + protobuf
ios: ffi + protobuf
web: wasm + protobuf
hover: go + protobuf

https://pub.dev/packages/openpgp/versions/1.0.0-rc1

**Comment by joe-getcouragenow** [comment]
Hey @jerson nice stuff !!!! 

I will try it out and feedback and also try to help on the code.

**Comment by jerson** [comment]
@joe-getcouragenow  thank you, im adding github actions to go repo, so now all binaries are generated on each tag, this should help to add features https://github.com/jerson/openpgp-mobile/releases/tag/v0.9.7

**Comment by joe-getcouragenow** [comment]
ok cool.
